File: leetcode/LinkedListTests/LinkedList_234.cs
namespace LinkedListTests;

[TestFixture]
class LinkedList_234
{
    class Solution {
        public bool IsPalindrome(ListNode head) {
            var halfHead = EndOfFirstHalf(head).next;
            var reverseHalf = ReverseList(halfHead);
            while(reverseHalf is not null) {
                if(reverseHalf.val != head.val) {
                    return false;
                }
                head = head.next;
                reverseHalf = reverseHalf.next;
            }
            return true;
        }

        private ListNode ReverseList(ListNode head) {
            ListNode newHead = null;
            ListNode beforeNewHead = null;
            while (head is not null) {
                beforeNewHead = new ListNode(head.val);
                beforeNewHead.next = newHead;
                newHead = beforeNewHead;
                head = head.next;
            }
            return newHead;
        }

        private ListNode EndOfFirstHalf(ListNode head) {
            ListNode fast = head;
            ListNode slow = head;
            while (fast.next != null && fast.next.next != null) {
                fast = fast.next.next;
                slow = slow.next;
            }
            return slow;
        }
    }
}

File: leetcodePython/LinkedList/linkedlist_234.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.next = next
        self.val = val


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        middle = fast = head
        # A -> B -> None return A
        # A -> B -> C -> None return B
        # fast.next.next not None make sure we have at least 3 element to move
        # A -> B -> None return A
        while fast and fast.next and fast.next.next:
            fast = fast.next.next
            middle = middle.next

        # A -------- middle -> ........end
        prev = None
        curr = middle.next
        while curr:
            curr_next = curr.next
            curr.next = prev
            prev = curr
            curr = curr_next

        # A -------- middle  |  after_middle <- ........prev
        half_head = prev
        while half_head:
            if half_head.val != head.val:
                return False
            head = head.next
            half_head = half_head.next
        return True
